### PR TITLE
ARROW-6578: [C++] Allow casting number to string

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@ursabot build

**Comment by pitrou** [comment]
I added a benchmark. Numbers here:
```
----------------------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations UserCounters...
----------------------------------------------------------------------------------------
IntegerFormatting<Int8Type>         4815 ns         4809 ns       581635 items_per_second=207.959M/s
IntegerFormatting<Int16Type>        6690 ns         6680 ns       419664 items_per_second=149.69M/s
IntegerFormatting<Int32Type>       12819 ns        12802 ns       218800 items_per_second=78.1118M/s
IntegerFormatting<Int64Type>       23434 ns        23403 ns       119849 items_per_second=42.729M/s
IntegerFormatting<UInt8Type>        2738 ns         2738 ns      1025531 items_per_second=365.284M/s
IntegerFormatting<UInt16Type>       6559 ns         6559 ns       426743 items_per_second=152.468M/s
IntegerFormatting<UInt32Type>      11191 ns        11191 ns       250162 items_per_second=89.3588M/s
IntegerFormatting<UInt64Type>      24299 ns        24298 ns       114985 items_per_second=41.1551M/s
FloatFormatting<FloatType>        134871 ns       134869 ns        20760 items_per_second=7.41459M/s
FloatFormatting<DoubleType>       136527 ns       136525 ns        20511 items_per_second=7.32465M/s
```

**Comment by pitrou** [comment]
@bkietz

File: cpp/src/arrow/util/formatting.h [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+ public:
+  explicit StringFormatter(const std::shared_ptr<DataType>& = NULLPTR) {}
+
+  using value_type = bool;
+
+  template <typename Appender>
+  Status operator()(bool value, Appender&& append) {
+    if (value) {
+      const char string[] = "true";
+      return append(string, static_cast<int32_t>(sizeof(string) - 1));
```
Comment (bkietz) [inline cpp/src/arrow/util/formatting.h]: Why not just use string_view?

File: cpp/src/arrow/util/formatting.h [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+  }
+};
+
+/////////////////////////////////////////////////////////////////////////
+// Integer formatting
+
+namespace detail {
+
+// A 2x100 direct table mapping integers in [0..99] to their decimal representations.
+ARROW_EXPORT extern const char digit_pairs[];
```
Comment (bkietz) [inline cpp/src/arrow/util/formatting.h]: Interesting. Out of curiosity, how much performance does this buy you?

**Review by bkietz — APPROVED** [review]

File: cpp/src/arrow/util/formatting.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/formatting.h]: It's around 30% slower if computing one digit at a time.

File: cpp/src/arrow/util/formatting.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/util/formatting.h]: Hmm, that's a good point. Will try.

**Comment by wesm** [comment]
Appveyor: https://ci.appveyor.com/project/pitrou/arrow/builds/27862731

**Pull request closed without merging**